src/box/box.h

```diff
@@ -70,6 +70,47 @@ extern double on_shutdown_trigger_timeout;
 /** Invoked on box shutdown. */
 extern struct rlist box_on_shutdown_trigger_list;
 
+/**
+ * Triggers invoked during initial `box.cfg` call on various recovery stages.
+ */
+extern struct rlist box_on_recovery_state;
+
+/**
+ * Recovery states supported by on_recovery_state triggers.
+ * Are positioned in the order of appearance during initial box.cfg().
+ * The only exception are WAL_RECOVERED and INDEXES_BUILT, which might come in
+ * any order, since the moment when secondary indexes are built depends on
+ * box.cfg.force_recovery and box.cfg.memtx_use_mvcc_engine.
+ */
+enum box_recovery_state {
+	/**
+	 * The node has either recovered the snapshot from the disk, received
+	 * the snapshot from the remote master as part of initial join stage or
+	 * has bootstrapped the cluster.
+	 */
+	RECOVERY_STATE_SNAPSHOT_RECOVERED,
+	/**
+	 * The node has either recovered the local WAL files, received the WALs
+	 * from the remote master as part of final join or has bootstrapped the
+	 * cluster.
+	 */
+	RECOVERY_STATE_WAL_RECOVERED,
+	/**
+	 * The node has built secondary indexes for memtx spaces.
+	 */
+	RECOVERY_STATE_INDEXES_BUILT,
+	/**
+	 * The node has synced with remote peers. IOW, it has transitioned from
+	 * "orphan" to "running".
+	 */
+	RECOVERY_STATE_SYNCED,
+	RECOVERY_STATE_MAX,
```

Comment:
Nit: For `STR2ENUM` and `ENUM2STR` helpers to work, this should be called `box_recovery_state_MAX`.